### PR TITLE
Make services tests nondestructive

### PR DESCRIPTION
**Pull request by marusak**
Builds on top of #13751 so blocked until that lands.

  - [x] Fix reboot hang after iSCSI tests: PR #13793

### CODE REVIEW

**Comment by marusak** [comment]
[32](https://logs-https-cockpit.apps.ci.centos.org/logs/pull-13758-20200320-111502-0e7321b7-fedora-32/log.html#266-2) and [31](https://logs.cockpit-project.org/logs/pull-13758-20200320-111502-0e7321b7-fedora-31-firefox/log.html#266-2) are not ideal.

**Comment by marusak** [comment]
Maybe I found out what the flake was. And it also would fix rather common flake with check-services tests, so :crossed_fingers:

File: test/common/testlib.py [hunk 1]
```diff
@@ -1271,6 +1271,17 @@ def restore_file(self, path, post_restore_action=None):
...
+    def write_file(self, path, content):
+        '''Write a new file on primary machine
+
+        This is safe for @nondestructive tests, the file will be removed during cleanup.
+        '''
+        m = self.machine
+        m.write(path, content)
+
+        if self.is_nondestructive():
+            self.addCleanup(m.execute, "rm {0}".format(path))
```
Comment (martinpitt) [inline test/common/testlib.py]: We may need to add `-f` at some point, when we use it in tests that remove the file. But we can do that if/when needed.

File: test/verify/check-services [hunk 2]
```diff
@@ -32,6 +33,15 @@ class TestServices(MachineCase):
         # we don't race with the boot process
         self.machine.execute("systemctl start default.target")
 
+        # We manipulate with `/etc/systemd/system` so `reload-daemon` to keep it in proper state
```
Comment (martinpitt) [inline test/verify/check-services]: "daemon-reload", not "reload-daemon".

File: test/verify/check-services [hunk 3]
```diff
@@ -32,6 +33,15 @@ class TestServices(MachineCase):
...
+        # Also `reset-failed` as failed units can be still listed with `systemctl` even when all files
+        # were removed and daemon reloaded.
+        self.restore_dir("/etc/systemd/system", "systemctl daemon-reload && systemctl reset-failed", True)
+
+        # Many services use this command and when the test finishes/fails while they are active,
+        # `systemct status test.service` still shows this service as active until this process dies
+        self.addCleanup(self.machine.execute, "pkill -fx '/bin/sh -c while true; do sleep 5; done' || true")
```
Comment (martinpitt) [inline test/verify/check-services]: Wouldn't it be more robust to `systemctl stop test.service`? Or are the service names different? Could it do something like `systemctl stop test*.service`?

File: test/verify/check-services [hunk 4]
```diff
@@ -135,6 +145,10 @@ Unit=test.service
         m.execute("systemctl daemon-reload")
         self.machine.execute("systemctl start default.target")
 
+        # When the test fails while `test.service` is active, this binary is kept running
+        # `systemct status test.service` still shows it as active until this process dies
+        self.addCleanup(self.machine.execute, "pkill -fx '/bin/sh /usr/local/bin/test-service' || true")
```
Comment (martinpitt) [inline test/verify/check-services]: Dito

File: test/verify/check-services (hunk 3 above)
Comment (marusak) [reply test/verify/check-services]: there are `test.service`, `condtest.service`, `test-a.service`, `test-b.service`, test-c.service`.
But those 3 are actually never started. So I can stop those specific services in appropriate tests.

**Comment by marusak** [comment]
[8.2](https://logs.cockpit-project.org/logs/pull-13758-20200324-100044-cc0297ba-rhel-8-2/log.html) and [8.2-distro](https://logs.cockpit-project.org/logs/pull-13758-20200324-100044-cc0297ba-rhel-8-2-distropkg/log.html#254-2) still fail on `testCreateTimer`. Was not able to reproduce it locally yet. Also the logs are for some reason too quiet.

**Comment by marusak** [comment]
hmm, disabling all machines tests and it [worked](https://logs.cockpit-project.org/logs/pull-13758-20200325-070913-6ceec505-rhel-8-2/log.html) - disabling some tests, as it was timing out with `-vt`. So maybe some machine tests is breaking it? :thinking:

**Comment by marusak** [comment]
oh, so `testCreateTimer` reboots the Vm and it is stuck on `A start job is running for Login an…SCSI devices (7min 41s / no limit)`. Eventually it succeeded, but this seems very suspicious.
This tests takes around 120s on clean VM, if on the VM ran some test with iscsi disks then it takes around 1050s and it is very likely, it timeouts in bots.

**Comment by martinpitt** [comment]
@marusak: Ah, this makes sense! I also ran into this neverending iSCSI start job back then when I converted `testAddDisk()` and the other one that does iSCSI. This means that `prepareStorageDeviceOnISCSI()` didn't clean up properly (or an additional thing that calls it)

**Comment by martinpitt** [comment]
I can reproduce the problem with running `TestMachinesDBus.testAddDisk` in an existing VM and then rebooting. Before/after the test,

 * there is no significant change in /etc
 * in /var the most significant change is in /var/lib/iscsi/{static,nodes} (but `targetcli ls` does not show any of these!); there is also some difference in /var/lib/nfs/
 * systemd; icsid.service and libvirtd.service and some NFS stuff is running

iscsi.service has `ConditionDirectoryNotEmpty=/var/lib/iscsi/nodes` which matches the unclean directory above. When I `rm -r /var/lib/iscsi`, the VM reboots just fine. So that gives us a way out:

```diff
--- test/verify/machineslib.py
+++ test/verify/machineslib.py
@@ -272,6 +272,8 @@ class TestMachines(MachineCase, StorageHelpers, NetworkHelpers):
         # Increase the iSCSI timeouts for heavy load during our testing
         self.sed_file(r"s|^\(node\..*log.*_timeout = \).*|\1 60|", "/etc/iscsi/iscsid.conf")
 
+        self.restore_dir("/var/lib/iscsi")
+
         # Setup a iSCSI target
         m.execute("""
                   targetcli /backstores/ramdisk create test 50M

```

I confirmed that this fixes the reboot problem.

However, that should come with a bug reference, as that's really rather unexpected -- `targetcli delete` ought to clean these up. I filed https://bugzilla.redhat.com/show_bug.cgi?id=1817241 about this with a standalone reproducer.

With that reproducer this happens on Fedora 32 as well, so I'm not sure why we only see that hang on rhel-8-2 in CI. Maybe in F32 it just times out fast enough to not run into a timeout, but it will still significantly slow down the subsequent tests that reboot.

**Comment by martinpitt** [comment]
I filed PR #13793 to fix the reboot hang, and mark this PR as blocking on that.

**Comment by martinpitt** [comment]
Yay green! :green_heart: :tada:

**Comment by marusak** [comment]
what :confused: ?
why is every testStoragePoolsCreate failing with `wait_js_cond(ph_is_visible("#storage-pool-dialog-target li a:contains(cockpittest):not([disabled])")): #storage-pool-dialog-target li a:contains(cockpittest):not([disabled]) is ambigous`

This must be something broken on master, cannot see how this PR could affect it.

**Comment by martinpitt** [comment]
@marusak: Mea culpa, my fault! Fixed in PR #13796 . The original PR #13793 succeeded because TestNondestructive happened to run as almost the last test.

**Pull request merged**